pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/StringPredicateFilterOptimizer.java

```diff
@@ -0,0 +1,127 @@
+  private static final String COMPARE_OPERATOR_NAME = "STRCMP";
+
+  @Override
+  public void optimize(PinotQuery query, @Nullable TableConfig tableConfig, @Nullable Schema schema) {
+    Expression filter = query.getFilterExpression();
+    if (filter != null && schema != null) {
+      optimizeExpression(filter, schema);
+    }
+
+    Expression expression = query.getHavingExpression();
+    if (expression != null && schema != null) {
+      optimizeExpression(expression, schema);
+    }
+  }
+
+  /** Traverse an expression tree to replace MINUS function with COMPARE if function operands are STRING. */
+  private static void optimizeExpression(Expression expression, Schema schema) {
+    ExpressionType type = expression.getType();
+    if (type != ExpressionType.FUNCTION) {
+      // We have nothing to rewrite if expression is not a function.
+      return;
+    }
+
+    Function function = expression.getFunctionCall();
+    String operator = function.getOperator();
+    List<Expression> operands = function.getOperands();
+    FilterKind kind = FilterKind.valueOf(operator);
+    switch (kind) {
+      case AND:
+      case OR: {
+        optimizeExpression(operands.get(0), schema);
+        break;
+      }
+      case EQUALS:
+      case NOT_EQUALS:
+      case GREATER_THAN:
+      case GREATER_THAN_OR_EQUAL:
+      case LESS_THAN:
+      case LESS_THAN_OR_EQUAL:
+      case IS_NULL:
+      case IS_NOT_NULL: {
+        replaceMinusWithCompareForStrings(operands.get(0), schema);
+        break;
+      }
+      default:
+        break;
+    }
+  }
+
+  /** Replace the operator of a MINUS function with COMPARE if both operands are STRING. */
+  private static void replaceMinusWithCompareForStrings(Expression expression, Schema schema) {
+    if (expression.getType() != ExpressionType.FUNCTION) {
+      // We have nothing to rewrite if expression is not a function.
+      return;
+    }
+
+    Function function = expression.getFunctionCall();
+    String operator = function.getOperator();
+    List<Expression> operands = function.getOperands();
+    if (operator.equals(MINUS_OPERATOR_NAME) && operands.size() == 2 && isStringColumn(operands.get(0), schema)
+        && isStringColumn(operands.get(1), schema)) {
+      function.setOperator(COMPARE_OPERATOR_NAME);
+    }
+  }
+
+  /** @return true if expression is a column of numeric type. */
+  private static boolean isStringColumn(Expression expression, Schema schema) {
+    if (expression.getType() != ExpressionType.IDENTIFIER) {
+      // Expression can not be a column.
+      return false;
+    }
+
+    String column = expression.getIdentifier().getName();
+    FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+    if (fieldSpec == null || !fieldSpec.isSingleValueField()) {
+      // Expression can not be a column name.
+      return false;
+    }
+
+    return schema.getFieldSpecFor(column).getDataType() == FieldSpec.DataType.STRING;
```

Comment:
Avoid fetching field spec twice
```suggestion
    return fieldSpec != null && fieldSpec.getDataType() == FieldSpec.DataType.STRING;
```

Reply:
I don't think we need to check single-value. IIRC, we will pass the first element of the MV to the function

Reply:
Refactored code a bit.